Feat/feature screen one unit tests

anarahari opened the pull request:
This branch consists of test cases for Screen one

Class on which test cases are performed
1. ViewModel - presentation layer
2. Use Case - domain layer

Test cases includes testing list of characters for success and failure

anarahari committed:
feat : Feature screen one, domain module setup

anarahari committed:
Merge branch 'feat/Workflow_EmptyModules' into feat/Feature_Screen_One_Domain

# Conflicts:
#	common/build.gradle.kts

anarahari committed:
feat : use case, repository, pagination setup

anarahari committed:
Merge branch 'feat/Workflow_EmptyModules' into feat/Feature_Screen_One_Domain

anarahari committed:
feat : setup data layer for screen one

anarahari committed:
Merge branch 'feat/Feature_Screen_One_Domain' into feat/Feature_Screen_One_Data

# Conflicts:
#	domain/src/main/java/com/compose/domain/paging/CharacterPagingSource.kt
#	domain/src/main/java/com/compose/domain/repository/CharacterRepository.kt
#	domain/src/main/java/com/compose/domain/usecase/GetCharactersUseCase.kt

anarahari committed:
feat : screen one data module setup with graphql, repo impl, dagger modules

anarahari committed:
feat : screen one presentation module setup

anarahari committed:
feat : top app bar for screes

anarahari committed:
feat : top app bar for screens

anarahari committed:
Merge branch 'feat/Feature_Screen_One_Domain' into feat/Feature_Screen_One_Data

anarahari committed:
feat : Added datamapper, repositoryimpl changes to integrate data mapper for character model

anarahari committed:
feat : ktlintcheck changes for spaces

anarahari committed:
feat : CommonConstants marked as internal

anarahari committed:
Merge branch 'feat/Feature_Screen_One_Data' into feat/Feature_Screen_One_Presentation

# Conflicts:
#	.idea/deploymentTargetSelector.xml
#	domain/build.gradle.kts

anarahari committed:
feat : FirstScreen topbar, listview using LazyVerticalGrid

anarahari committed:
feat : TopAppBar changes, removed preview function

anarahari committed:
feat : Removed flow from repository impl, added dispatchermodule, removed unwanted codes, rename typography to RMTypography, backactio renamed to onBackButtonPressed

anarahari committed:
feat : dispatcher module injection added in ViewModelModule, ViewModelFactoryProvider, Datamodule

anarahari committed:
feat : ktlint checks

anarahari committed:
feat : Test case for viewmodel

Test case to get list of characters on success for viewmodel

anarahari committed:
feat : Test case for viewmodel

- Test case to get list of characters on success
- Test case to show error while fetching characters list
- Test case to check for initial loading stage while    fetching characters list

anarahari committed:
feat : Test case for usecase

- Test case to get list of characters on success
- Test case to show error while fetching characters list

anarahari committed:
feat : ktlint issues

- Solved space issues in CharacterViewModelTest.kt, CharacterTestRepository.kt, GetCharactersUseCaseTest.kt

anarahari committed:
feat : data

- CharacterPagingSource removed commented code

anarahari committed:
Merge branch 'feat/Feature_Screen_One_Domain' into feat/Feature_Screen_One_Data

anarahari committed:
Merge branch 'feat/Feature_Screen_One_Domain' into feat/Feature_Screen_One_Data

anarahari committed:
Merge branch 'feat/Feature_Screen_One_Domain' into feat/Feature_Screen_One_Data

anarahari committed:
feat : data

- Changed the implementation type of getCharacters to   List<T> in repository
- Nullable return of charactermodel in datamapper

anarahari committed:
Merge branch 'feat/Feature_Screen_One_Data' into feat/Feature_Screen_One_Presentation

# Conflicts:
#	data/src/main/java/com/compose/data/paging/CharacterPagingSource.kt
#	domain/src/main/java/com/compose/domain/usecase/GetCharactersUseCase.kt

anarahari committed:
feat : presentation

- changes to calling of invoke function in usecase and viewmodel

anarahari committed:
Merge branch 'feat/Feature_Screen_One_Presentation' into feat/Feature_Screen_One_UnitTests

# Conflicts:
#	presentation/src/main/java/com/compose/presentation/viewmodel/CharacterViewModel.kt

anarahari committed:
feat : unittest

- added RelaxedMockK to characterusecase in viewmodeltest

shubhamsinghmutualmobile commented on domain/src/main/java/com/compose/domain/usecase/GetCharactersUseCase.kt:
```diff
@@ -1,12 +1,26 @@
-class GetCharactersUseCase @Inject constructor(private val characterRepository: CharacterRepository) {
+class GetCharactersUseCase @Inject constructor(
+    private val characterRepository: CharacterRepository,
+    @IoDispatcher private val ioDispatcher: CoroutineDispatcher
```
A better name for this variable would be `coroutineDispatcher`. The body of this class doesn't care whether the dispatcher is `IO`, `Main`, or `Default`. It only cares whether it's a dispatcher or not.

anarahari replied:
This is done

shubhamsinghmutualmobile commented on domain/src/test/java/com/compose/domain/usecase/GetCharactersUseCaseTest.kt:
```diff
@@ -0,0 +1,53 @@
+@ExperimentalCoroutinesApi
+class GetCharactersUseCaseTest {
+
+    private lateinit var charactersUseCase: GetCharactersUseCase
```
The class we test should be named `sut` (short for `System Under Test`) instead

shubhamsinghmutualmobile commented on domain/src/test/java/com/compose/domain/usecase/GetCharactersUseCaseTest.kt:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    @Test
+    fun `get list of characters on failure`() = runTest(testDispatcher) {
```
The name of a test should describe the expected behaviour of the function. Right now it says that we should expect a list of characters even for failure cases, which is not true.

So please try to name your test functions to depict their expected behaviour more accurately. Maybe follow a naming convention if that helps you.

You can use this as a reference: https://betterprogramming.pub/android-unit-testing-choosing-naming-convention-and-test-cases-d1a3122ac28a

anarahari replied:
ok sure, thanks for the feedback and link, i will read the document and understand the naming conventions and implement the same

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/screens/TopAppBar.kt:
```diff
@@ -0,0 +1,52 @@
+        onBackButtonPressed?.let {
+            Box(
+                modifier = Modifier
+                    .clickable { onBackButtonPressed }
```
Does this even work? Have you verified it? 🤔 In my understanding, it should not, since the lambda hasn't been invoked.

anarahari replied:
Yes the TopAppBar function works, as i have passed string title and null from the FirstScreen
I have made this function for future use as in second screen there will be a back button to come to the first screen

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/viewmodel/ViewModelFactoryProvider.kt:
```diff
@@ -0,0 +1,23 @@
+class ViewModelFactoryProvider
+@Inject constructor(
+    private val getCharactersUseCase: GetCharactersUseCase,
+    @IoDispatcher private val ioDispatcher: CoroutineDispatcher
```
Same, rename this to `coroutineDispatcher` only. Only your DI framework should know which specific type to inject, others should only be concerned with the generic type.

anarahari replied:
done

anarahari committed:
feat : Presentation

Pr reviews changes

- Changed naming convention for ioDispatcher to coroutineDispatcher

anarahari committed:
Merge branch 'feat/Feature_Screen_One_Presentation' into feat/Feature_Screen_One_UnitTests

anarahari committed:
feat : Presentation

Pr reviews changes

- Changed the class name to test as SUT (System under test)
- Changed the function name as per expected behaviour
